File: research/experiments/sweep/runs/opt-small-go-sonnet45/nelder_mead.go
package optimization

import (
	"math"
	"sort"
)

// NelderMeadOptions contains Nelder-Mead specific parameters
type NelderMeadOptions struct {
	OptimizeOptions
	Alpha              float64 // Reflection coefficient
	Gamma              float64 // Expansion coefficient
	Rho                float64 // Contraction coefficient
	Sigma              float64 // Shrink coefficient
	InitialSimplexScale float64 // Edge length scale
}

// DefaultNelderMeadOptions returns default Nelder-Mead options
func DefaultNelderMeadOptions(overrides *NelderMeadOptions) NelderMeadOptions {
	opts := NelderMeadOptions{
		OptimizeOptions:     DefaultOptions(nil),
		Alpha:               1.0,
		Gamma:               2.0,
		Rho:                 0.5,
		Sigma:               0.5,
		InitialSimplexScale: 0.05,
	}

	if overrides != nil {
		if overrides.GradTol != 0 {
			opts.GradTol = overrides.GradTol
		}
		if overrides.StepTol != 0 {
			opts.StepTol = overrides.StepTol
		}
		if overrides.FuncTol != 0 {
			opts.FuncTol = overrides.FuncTol
		}
		if overrides.MaxIterations != 0 {
			opts.MaxIterations = overrides.MaxIterations
		}
		if overrides.Alpha != 0 {
			opts.Alpha = overrides.Alpha
		}
		if overrides.Gamma != 0 {
			opts.Gamma = overrides.Gamma
		}
		if overrides.Rho != 0 {
			opts.Rho = overrides.Rho
		}
		if overrides.Sigma != 0 {
			opts.Sigma = overrides.Sigma
		}
		if overrides.InitialSimplexScale != 0 {
			opts.InitialSimplexScale = overrides.InitialSimplexScale
		}
	}

	return opts
}

// NelderMead minimizes a function using the Nelder-Mead simplex method
func NelderMead(f func([]float64) float64, x0 []float64, options *NelderMeadOptions) OptimizeResult {
	opts := DefaultNelderMeadOptions(options)
	n := len(x0)

	// Create initial simplex
	simplex, fValues := createInitialSimplex(f, x0, opts.InitialSimplexScale)
	functionCalls := n + 1

	iteration := 0
	for {
		// Sort simplex by function values
		indices := make([]int, len(fValues))
		for i := range indices {
			indices[i] = i
		}
		sort.Slice(indices, func(i, j int) bool {
			return fValues[indices[i]] < fValues[indices[j]]
		})

		// Reorder simplex and fValues
		sortedSimplex := make([][]float64, len(simplex))
		sortedFValues := make([]float64, len(fValues))
		for i, idx := range indices {
			sortedSimplex[i] = simplex[idx]
			sortedFValues[i] = fValues[idx]
		}
		simplex = sortedSimplex
		fValues = sortedFValues

		// Check convergence
		reason := checkNelderMeadConvergence(simplex, fValues, iteration, opts)
		if reason != nil {
			return OptimizeResult{
				X:             Clone(simplex[0]),
				Fun:           fValues[0],
				Gradient:      nil,
				Iterations:    iteration,
				FunctionCalls: functionCalls,
				GradientCalls: 0,
				Converged:     IsConverged(reason),
				Message:       ConvergenceMessage(reason),
			}
		}

		// Check max iterations
		if iteration >= opts.MaxIterations {
			return OptimizeResult{
				X:             Clone(simplex[0]),
				Fun:           fValues[0],
				Gradient:      nil,
				Iterations:    iteration,
				FunctionCalls: functionCalls,
				GradientCalls: 0,
				Converged:     false,
				Message:       "maximum iterations reached",
			}
		}

		// Compute centroid of all vertices except worst
		centroid := Zeros(n)
		for i := 0; i < n; i++ {
			centroid = Add(centroid, simplex[i])
		}
		centroid = Scale(centroid, 1.0/float64(n))

		// Reflection
		reflected := AddScaled(centroid, Sub(centroid, simplex[n]), opts.Alpha)
		fReflected := f(reflected)
		functionCalls++

		// Accept reflection if between best and second-worst
		if fReflected >= fValues[0] && fReflected < fValues[n-1] {
			simplex[n] = reflected
			fValues[n] = fReflected
			iteration++
			continue
		}

		// Expansion (if reflection is best)
		if fReflected < fValues[0] {
			expanded := AddScaled(centroid, Sub(reflected, centroid), opts.Gamma)
			fExpanded := f(expanded)
			functionCalls++

			if fExpanded < fReflected {
				simplex[n] = expanded
				fValues[n] = fExpanded
			} else {
				simplex[n] = reflected
				fValues[n] = fReflected
			}
			iteration++
			continue
		}

		// Contraction
		var contracted []float64
		var fContracted float64

		if fReflected < fValues[n] {
			// Outside contraction
			contracted = AddScaled(centroid, Sub(reflected, centroid), opts.Rho)
			fContracted = f(contracted)
			functionCalls++

			if fContracted <= fReflected {
				simplex[n] = contracted
				fValues[n] = fContracted
				iteration++
				continue
			}
		} else {
			// Inside contraction
			contracted = AddScaled(centroid, Sub(simplex[n], centroid), opts.Rho)
			fContracted = f(contracted)
			functionCalls++

			if fContracted < fValues[n] {
				simplex[n] = contracted
				fValues[n] = fContracted
				iteration++
				continue
			}
		}

		// Shrink
		for i := 1; i <= n; i++ {
			simplex[i] = AddScaled(simplex[0], Sub(simplex[i], simplex[0]), opts.Sigma)
			fValues[i] = f(simplex[i])
		}
		functionCalls += n
		iteration++
	}
}

// createInitialSimplex creates the initial simplex for Nelder-Mead
func createInitialSimplex(f func([]float64) float64, x0 []float64, scale float64) ([][]float64, []float64) {
	n := len(x0)
	simplex := make([][]float64, n+1)
	fValues := make([]float64, n+1)

	// First vertex is x0
	simplex[0] = Clone(x0)
	fValues[0] = f(x0)

	// Remaining vertices: x0 + h*ei
	for i := 0; i < n; i++ {
		h := scale * math.Max(math.Abs(x0[i]), 1.0)
		vertex := Clone(x0)
		vertex[i] += h
		simplex[i+1] = vertex
		fValues[i+1] = f(vertex)
	}

	return simplex, fValues
}

// checkNelderMeadConvergence checks convergence based on simplex spread and diameter
func checkNelderMeadConvergence(simplex [][]float64, fValues []float64, iteration int, opts NelderMeadOptions) *ConvergenceReason {
	n := len(simplex[0])

	// Function value spread (standard deviation)
	mean := 0.0
	for _, fv := range fValues {
		mean += fv
	}
	mean /= float64(len(fValues))

	variance := 0.0
	for _, fv := range fValues {
		diff := fv - mean
		variance += diff * diff
	}
	variance /= float64(len(fValues))
	stdDev := math.Sqrt(variance)

	if stdDev < opts.FuncTol {
		return &ConvergenceReason{Kind: "function"}
	}

	// Simplex diameter (max distance from best vertex)
	maxDist := 0.0
	for i := 1; i <= n; i++ {
		dist := NormInf(Sub(simplex[i], simplex[0]))
		if dist > maxDist {
			maxDist = dist
		}
	}

	if maxDist < opts.StepTol {
		return &ConvergenceReason{Kind: "step"}
	}

	// No convergence
	return nil
}

File: research/experiments/sweep/runs/opt-small-cpp-sonnet45/src/nelder_mead.cpp
#include "nelder_mead.h"
#include <algorithm>
#include <cmath>
#include <numeric>

namespace optimization {

namespace {
    // Nelder-Mead parameters (universal across scipy, MATLAB, Optim.jl)
    const double ALPHA = 1.0;   // Reflection
    const double GAMMA = 2.0;   // Expansion
    const double RHO = 0.5;     // Contraction
    const double SIGMA = 0.5;   // Shrink
    const double INITIAL_SIMPLEX_SCALE = 0.05;

    struct Vertex {
        Vector point;
        double value;

        bool operator<(const Vertex& other) const {
            return value < other.value;
        }
    };

    // Create initial simplex: vertex 0 = x0, vertex i = x0 + h*e_i
    std::vector<Vertex> createInitialSimplex(const ObjectiveFunction& f, const Vector& x0) {
        size_t n = x0.size();
        std::vector<Vertex> simplex(n + 1);

        // First vertex is the starting point
        simplex[0].point = x0;
        simplex[0].value = f(x0);

        // Subsequent vertices offset along each dimension
        for (size_t i = 0; i < n; ++i) {
            Vector point = clone(x0);
            double h = INITIAL_SIMPLEX_SCALE * std::max(std::abs(x0[i]), 1.0);
            point[i] += h;
            simplex[i + 1].point = point;
            simplex[i + 1].value = f(point);
        }

        return simplex;
    }

    // Compute centroid of all vertices except the worst
    Vector computeCentroid(const std::vector<Vertex>& simplex) {
        size_t n = simplex[0].point.size();
        Vector centroid = zeros(n);

        // Sum all vertices except the last (worst)
        for (size_t i = 0; i < simplex.size() - 1; ++i) {
            centroid = add(centroid, simplex[i].point);
        }

        // Divide by count
        return scale(centroid, 1.0 / (simplex.size() - 1));
    }

    // Compute simplex diameter (maximum distance between any two vertices)
    double simplexDiameter(const std::vector<Vertex>& simplex) {
        double maxDist = 0.0;
        for (size_t i = 0; i < simplex.size(); ++i) {
            for (size_t j = i + 1; j < simplex.size(); ++j) {
                double dist = norm(sub(simplex[i].point, simplex[j].point));
                maxDist = std::max(maxDist, dist);
            }
        }
        return maxDist;
    }

    // Compute standard deviation of function values
    double functionSpread(const std::vector<Vertex>& simplex) {
        double mean = 0.0;
        for (const auto& v : simplex) {
            mean += v.value;
        }
        mean /= simplex.size();

        double variance = 0.0;
        for (const auto& v : simplex) {
            double diff = v.value - mean;
            variance += diff * diff;
        }
        variance /= simplex.size();

        return std::sqrt(variance);
    }
}

OptimizeResult nelderMead(
    const ObjectiveFunction& f,
    const Vector& x0,
    const OptimizeOptions& options
) {
    // Initialize simplex
    std::vector<Vertex> simplex = createInitialSimplex(f, x0);
    int functionCalls = static_cast<int>(simplex.size());
    int iterations = 0;

    while (iterations < options.maxIterations) {
        // Sort simplex by function value (ascending)
        std::sort(simplex.begin(), simplex.end());

        const Vertex& best = simplex[0];
        const Vertex& secondWorst = simplex[simplex.size() - 2];
        const Vertex& worst = simplex[simplex.size() - 1];

        // Check convergence: function spread or simplex diameter
        double spread = functionSpread(simplex);
        double diameter = simplexDiameter(simplex);

        if (spread < options.funcTol || diameter < options.stepTol) {
            return OptimizeResult{
                best.point,
                best.value,
                std::nullopt,  // gradient
                iterations,
                functionCalls,
                0,  // gradientCalls
                true,
                "Converged: simplex tolerance met"
            };
        }

        // Compute centroid of all but worst
        Vector centroid = computeCentroid(simplex);

        // Reflection: xr = centroid + alpha * (centroid - worst)
        Vector reflected = addScaled(centroid, sub(centroid, worst.point), ALPHA);
        double fReflected = f(reflected);
        functionCalls++;

        // Accept reflection if between best and second-worst
        if (fReflected >= best.value && fReflected < secondWorst.value) {
            simplex.back().point = reflected;
            simplex.back().value = fReflected;
            iterations++;
            continue;
        }

        // If reflection is best, try expansion
        if (fReflected < best.value) {
            Vector expanded = addScaled(centroid, sub(reflected, centroid), GAMMA);
            double fExpanded = f(expanded);
            functionCalls++;

            if (fExpanded < fReflected) {
                simplex.back().point = expanded;
                simplex.back().value = fExpanded;
            } else {
                simplex.back().point = reflected;
                simplex.back().value = fReflected;
            }
            iterations++;
            continue;
        }

        // Reflection is worst or second-worst, try contraction
        Vector contracted;
        double fContracted;

        if (fReflected < worst.value) {
            // Outside contraction: between centroid and reflected
            contracted = addScaled(centroid, sub(reflected, centroid), RHO);
            fContracted = f(contracted);
            functionCalls++;

            if (fContracted <= fReflected) {
                simplex.back().point = contracted;
                simplex.back().value = fContracted;
                iterations++;
                continue;
            }
        } else {
            // Inside contraction: between centroid and worst
            contracted = addScaled(centroid, sub(worst.point, centroid), RHO);
            fContracted = f(contracted);
            functionCalls++;

            if (fContracted < worst.value) {
                simplex.back().point = contracted;
                simplex.back().value = fContracted;
                iterations++;
                continue;
            }
        }

        // Contraction failed, shrink all vertices toward best
        for (size_t i = 1; i < simplex.size(); ++i) {
            simplex[i].point = addScaled(best.point, sub(simplex[i].point, best.point), SIGMA);
            simplex[i].value = f(simplex[i].point);
            functionCalls++;
        }

        iterations++;
    }

    // Max iterations reached
    std::sort(simplex.begin(), simplex.end());
    return OptimizeResult{
        simplex[0].point,
        simplex[0].value,
        std::nullopt,  // gradient
        iterations,
        functionCalls,
        0,  // gradientCalls
        false,
        "Maximum iterations reached"
    };
}

} // namespace optimization
